qmxgraph/html.py

```diff
@@ -1,22 +1,108 @@
+                                     value=value, i=i)
+                    raise TypeError(msg)
+        else:
+            msg = "'{name}' must be a tuple but got {value!r}"
+            raise TypeError(msg.format(name=attr.name, value=value))
+
+
+def content_validator(child_types):
     """
-    Create an image tag that could be embedded into a table contents.
+    Creates an validator that accept an item with any type listed or a tuple
+    with any of the accept types combination). A tuple is used due it's
+    immutable nature.
 
-    The image's width and height and required since mxgraph will render the
+    :param list[type] child_types:
+    :rtype: Callable
+    :return: An callable to be used as an validator with the `attr` module.
+    ..see: http://www.attrs.org/en/stable/examples.html#callables
+    """
+    return _HtmlChildTypeValidator(child_types=tuple(child_types))
+
+
+@attr.s(frozen=True, slots=True)
+class Image:
+    """
+    Represet an image tag that could be embedded into a table contents.
+
+    The image's width and height are required since mxgraph will render the
     html in a helper container in order to get the cell's size. To avoid the
     cell size to be wrongly calculated we got some options like passing the
     image's size explicitly (as done here) or force the user to pre load the
     images so when rendering the html the image is already loaded and the
     correct size is used.
 
-    :param str source: The URL to the image, data URIs can be used.
-    :param int width: The desired width for the image.
-    :param int height: The desired height for the image.
-    :rtype: str
-    """
-    tag_template = '<img src="{source}" width="{width}" height="{height}" />'
-    if source.find('"') != -1:
-        raise ValueError('source argument can not have the " character')
-    width = int(width)
-    height = int(height)
-    return tag_template.format(source=source, width=width, height=height)
+    :ivar str tag:
+    :ivar str source: The URL to the image, data URIs can be used.
+    :ivar int width: The desired width for the image.
+    :ivar int height: The desired height for the image.
+    """
+    tag = attr.ib(default='img', init=False)
+    src = attr.ib(validator=attr.validators.instance_of(str))
+    width = attr.ib(validator=attr.validators.instance_of(int))
+    height = attr.ib(validator=attr.validators.instance_of(int))
+
+
+@attr.s(frozen=True, slots=True)
+class TableData:
+    """
+    Represents the contents of a table's cell when inserting or updating a
+    table in the graph.
+
+    :ivar str tag:
+    :ivar tuple[union[str,Image]] contents: The table cell's contents.
+    :ivar int colspan: The number of columns the cell should span into.
+    :ivar int rowspan: The number of rows the cell should span into.
+    """
+    tag = attr.ib(default='td', init=False)
+    contents = attr.ib(validator=content_validator([str, Image]),
+                       convert=tuple)
```

Comment:
`convert` is called before `validator`.
I also wrote the tuple constrain on validator before adding the `convert`. Since calling the constructors with lists seen more natural. Tuples are nice for they immutability but than the code looks awfully like lisp, no problems with lisp it self or even parenthesis.